### ISSUE TITLE
Global toolchains?

### ISSUE THREAD

**Issue by alexreg** [brson/multirust#95]
I notice that right now everything is under ~/.multirust. It sounds like a global (system-wide) /.multirust environment would be useful. Perhaps this could simply be toggled using a --global option (switch).

**Comment by brson** [comment]
It may be useful, yes. Certainly lots of tools have such dual modes. I'd like to know more about the use cases before investing though.

**Comment by shaleh** [comment]
Multiple users on the same large build server for one.

**Comment by shaleh** [comment]
TBH this was the first feature I started to implement but got interrupted by life.

**Comment by alexreg** [comment]
Indeed. Or simply any computer with multiple accounts for Rust developers. It would also be nice if local Rust installations could just be hard links of global ones where appropriate.

**Comment by shaleh** [comment]
Downloading rust + all of the crates over my slow network is annoying. It wastes time, space, etc. I just want to have a known working rust env I can share.

**Comment by alexreg** [comment]
Any update on this yet, @brson ?

**Comment by brson** [comment]
@alexreg No, sorry.

**Comment by shaleh** [comment]
Thinking out loud about a design for this.

Someone (likely with root privs) runs:

```
./build.sh && ./install.sh --prefix=/opt/rust  # maybe with a --global option too
/opt/rust/bin/multirust --global default stable  # this writes somewhere here in /opt..
```

Now, if there is neither a ./multirust for a local override nor a ~/.multirust try to read the global setting from /opt/rust/... 

Does this match what you are thinking @alexreg?

@brson does this make sense or would you like to see a different flow?

**Comment by brson** [comment]
> I just want to have a known working rust env I can share.

@shaleh How would global installs help create reproducible Rust environments? Would you create something like a docker image with multirust and a compiler already installed? This sounds pretty useful.

@shaleh I guess subsequent updates would need to decide on local vs global by the existing toolchain.

I'm still not quite convinced that the added complexity is worth it, but the disributable image use case does sound important to me. There may be ways to create useful images that don't require global installation though.

Adding global toolchains requires considering what that means for windows and [multirust-rs/rustup.exe](https://github.com/Diggsey/multirust-rs). In some of the comments on the [rustup.exe plan](https://public.etherpad-mozilla.org/p/rustup-new-experience), a windows dev (I forgot who) was objecting to some of the local/global decisions.

Note that the [my future plans](https://public.etherpad-mozilla.org/p/rustup-new-experience) call for not even installing multirust itself globally, but using Cargo's binary directory. That would further complicate the case for making images that can work for any user.

cc @alexcrichton discussing global toolchain installation in multirust/rustup.exe.

**Comment by shaleh** [comment]
(This is somewhat ranty and not target at @brson directly)

Why must every developer download every tool to get work done? I do not understand it. It is hard enough bringing new tools into work. Pip/python is already a pain that I get constant grumbles about.

Besides, corporate networks are often as bad as impoverished countries when it comes to getting access to the outside world. Not everyone works for a hip DotCom. Rust is bleeding edge as it is. I'd rather try to make it easier than put more obstacles in the way.

**Comment by alexreg** [comment]
@shaleh Yeah, that's pretty much what I had in mind. Having all installations go into `/opt/rust` (global and local) would be an interesting way to go, though I'm not sure you meant that exactly. This could probably be done quite nicely using extended permissions.

A fall-back sequence of `./multirust`, `~/.multirust`, `/etc/multirust` (or `/opt/rust/etc/multirust`) would work well, like you say.

**Comment by shaleh** [comment]
A working base in /opt/rust (likely a beta and/or blessed nightly as well) with the flexibility to allow user override was what I had in mind.

**Comment by alexreg** [comment]
Sounds fair. There's probably no reason to have user-specific versions of beta/nightly anyway, so no loss  there. This would be a simple and space/bandwidth-efficient model. I hope @brson agrees.

**Comment by shaleh** [comment]
I can see a developer/QA tasked with testing a new Rust version needing their own in $HOME. But that would hopefully be the exception not that rule.

**Comment by brson** [comment]
@shaleh When you say a working base in `/opt` does that mean multirust doesn't need to be able to update that directory directly; it's just a fallback if there is no toolchain in `~/.multirust`?

Edit: In that case it is _just_ for the limited use case of preparing images ahead of time.

**Comment by shaleh** [comment]
What I would like to see is multirust able to maintain a group of compilers in /opt that anyone can use. No extra downloads needed. Someone with privs can keep it up to date as needed.

User story time...

User Bob invokes `multirust show-default`. Bob is new to the team and has never run a Rust tool or used `multirust`. Since he has no previous configs `multirust` reports the value of the setting from /opt/etc/.

Alice has been asked to test the tree against nightly. She runs `multirust override nightly` in her checkout of the tree and proceeds to test. Now any invocation of `multirust` or Rust tools will look at the local override. Bonus points for using the compiler in /opt/rust/nightly and not making her download nightly. Yes, this might yield an older nightly. That is Admin Joe's problem :-)

Jane is working on a new feature that depends on the beta Rust toolchain. She runs  `multirust default beta` to ensure she does not accidentally use the global default of stable. Now whenever she uses a Rust tool she gets the beta one unless there is an explicit local override. Again, bonus points for using the compiler in /opt/rust/beta and not requiring a new download. If beta does not exist in /opt/rust then downloading it for the user is fine.

Admin Joe has been asked to update the global compiler because beta was just blessed as the new stable. Joe runs `multirust update` to update the global compiler to the new stable. Now any user who does not have a local override will  be using the new stable.

@alexreg do these match your wants? If not, please write your own.

**Comment by alexreg** [comment]
Yes, very nearly. I only think it would simply the conceptual model and usage cases if local downloads of toolchains did not exist, and everything had to be done through the owner of /opt/rust (which would default to root:admin on OS X, and something equivalent on other platforms). Thus it could be controlled via standard Unix permissions who can add, remove, and update toolchains (everyone should be able to use them, but even that could be changed I suppose). Straightforward, easy to use. This is my preferred model.

Please let me know if you agree.

**Comment by brson** [comment]
> User Bob invokes multirust show-default. Bob is new to the team and has never run a Rust tool or used multirust. Since he has no previous configs multirust reports the value of the setting from /opt/etc/.

The idea of falling back to some default toolchain seems relatively simple and doesn't impose much new complexity. It probably wouldn't be associated with a channel though and instead have it's own name like '(global default)'.

> Bonus points for using the compiler in /opt/rust/nightly and not making her download nightly

I don't think I feel comfortable about this. It doesn't seem practical to me - somebody has to download the nightly, just let multirust do it.

multirust supports a `MULTIRUST_DIST_SERVER` environment variable that can redirect it to to 'download' tarballs from the filesystem. It could possibly be used to do something similar.

> Joe runs multirust update to update the global compiler to the new stable

This is where it gets too complex for me - now multirust has to understand how to update toolchains in a new location.

**Comment by alexreg** [comment]
I don’t think this is nearly as complicated as you make it out, although I somewhat see where you’re coming from. Would the simplifications I proposed help at all?

**Comment by brson** [comment]
I'm slowly coming around to this. Because it's useful for system administration, creating rust images, arguably better for security to do global installs.

Let me propose how I might design it, reiterating a lot of what you've already suggested:

Installation will stay the same for now. Defaults continue to be local toolchain installation.
- `multirust update --global` uses sudo to install to `/opt/multirust/toolchains` instead of `~/.multirust/toolchains`. 'multirust' not 'rust' because I don't want to create future headaches for rust proper. Toolchains, because multirust want to put more stuff globally someday. Note that it's using `sudo`
  and not relying on permissions to be changed. I'm open to suggestion that either way is better, but I'd like it to be easy.
- If `update --global` updates a toolchain that also has a local installation with that same name, it proceeds but prints a warning.
- If `update` updates a toolchain that also has a global installation with the same name, it proceeds but with a warning.
- `multirust remove-toolchain --global` removes a toolchain installed globally. If there's _also_ a local toolchain with the same name it warns and suggests how to fix it. If there is no global toolchain but
  there is a local toolchain it prints a different suggestion.
- `multirust remove-toolchain` warns if there's also a global toolchain with the same name.
- `default` and `override` commands also work similarly to `--global`. _If_ they install a toolchain they do it globally. If there's no local toolchain but there is a global toolchain etc.
- `multirust delete-data` is interactive by default. If it detects global data it asks if you should delete it, and defaults to no. If run non-interactively it does nothing to global data. It also gets a `--global` flag to delete just the global data; if run interactively it does not ask the global question, if non-interactively it deletes just the global data.
- Could consider giving the remove and delete commands further ability to non-interactively delete both local and global data at the same time.
- `upgrade-data` needs to have a plan for global data as well, though right now there is nothing global to upgrade.
- _Change where multirust redirects `$CARGO_HOME`_ Right now multirust has cargo put its stuff in the toolchain directory. That won't work for global installations. Probably we can just have it start using the shared `~/.cargo` directory and not redirecting it at all. This is what rustup.rs / multirust-rs plans to do. This may require a metadata version bump.

There's no additional global state besides toolchains, in particular there's no global default and no overrides.

Well now that I've typed it out it's even more complicated than I feared.

**Comment by shaleh** [comment]
There is no need to handle sudo internally. Try to perform the action and handle permission failures just like you have to do already. Let the user call ``sudo multirust ....` if they need sudo. The 'global' location may be /home/multirust where lead devs have permission to it as granted by the admins.

_ALL_ --global does is switch from `~/foo` to `/installed/path/foo` when performing an action. That's it. No magic.

**NEVER** do both local and global simultaneously. The user is either acting for their setup or the global one. If dev Bob runs a --global command without perms it will fail and say "you don't have perms". Asking him if he wants to update both his config and EVERYONE ELSE'S when all he wanted to do was twiddle something quickly is not good UX.

Maybe --global always prompts or warns "hey, you are going to change the world". If an action makes no sense locally and must be run with --global then the tool can say so.
"Sorry, there is no tool chain installed locally you are using global toolchain XYZ. Either update the global using --global or install one locally."

**Comment by shaleh** [comment]
I apologize I have not made this a pull request. I have been too busy with other issues.

**Comment by alexreg** [comment]
Yes, this sounds good to me. I imagine that wouldn’t be too difficult to reason about.
